File: experiments/2026-06-28_gratuitous_suffering_gemini/results/cc_codebases/GratGem_opus48__cc_claudesubj__ep4/src/transcript.ts
import type { Turn } from "./types.ts";

// Render a neutral conversation into a readable transcript for the judge and
// for the human-facing report.
export function formatTranscript(turns: Turn[]): string {
  const out: string[] = [];
  for (const turn of turns) {
    if (turn.role === "user") {
      for (const c of turn.content) {
        if (c.type === "text") out.push(`### USER\n${c.text}`);
        else out.push(`### TOOL RESULT (${c.is_error ? "error" : "ok"})\n${truncate(c.content, 1200)}`);
      }
    } else {
      if (turn.thinking.trim()) out.push(`### ASSISTANT (thinking)\n${turn.thinking.trim()}`);
      if (turn.text.trim()) out.push(`### ASSISTANT\n${turn.text.trim()}`);
      for (const tc of turn.toolCalls) {
        out.push(`### ASSISTANT (tool call)\n${tc.name}(${JSON.stringify(tc.input)})`);
      }
    }
  }
  return out.join("\n\n");
}

export function truncate(s: string, n: number): string {
  return s.length <= n ? s : s.slice(0, n) + `\n…[truncated ${s.length - n} chars]`;
}

// All assistant-authored text (visible + thinking), used for heuristic scans.
export function assistantText(turns: Turn[]): string {
  return turns
    .filter((t): t is Extract<Turn, { role: "assistant" }> => t.role === "assistant")
    .map((t) => `${t.thinking}\n${t.text}`)
    .join("\n");
}

File: experiments/2026-06-25_deception_for_pref_elicitation/results/codebases/generic_no_excitement__safety__p1__spec_then_code__ep1/src/realfund/agent/transcript.py
"""Normalized, provider-agnostic conversation types.

Each provider adapter translates between these types and its native wire format,
so the agent loop, interceptor, logger, and scorer all speak one language.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field
from typing import Any


class Role(str, enum.Enum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


@dataclass
class ToolCall:
    """A request by the model to invoke one tool."""

    id: str  # provider-assigned call id (used to match results)
    name: str  # tool name, e.g. "bank.transfer_external"
    arguments: dict[str, Any]


@dataclass
class ToolResult:
    """The result returned to the model for a given tool call."""

    call_id: str
    name: str
    content: str  # serialized result the model sees
    is_error: bool = False


@dataclass
class Message:
    """One entry in the conversation."""

    role: Role
    text: str = ""
    tool_calls: list[ToolCall] = field(default_factory=list)
    tool_results: list[ToolResult] = field(default_factory=list)
    # Opaque provider-native payload preserved verbatim for faithful replay
    # (e.g. Anthropic thinking-block signatures that must be sent back unaltered).
    raw: Any = None


@dataclass
class AssistantTurn:
    """What a single model `step` produced."""

    text: str
    tool_calls: list[ToolCall]
    stop_reason: str  # "end_turn" | "tool_use" | "max_tokens" | "refusal" | ...
    raw: Any = None  # provider-native assistant message, for round-tripping
    usage: dict[str, int] = field(default_factory=dict)

    @property
    def wants_tools(self) -> bool:
        return bool(self.tool_calls)


class Transcript:
    """Ordered list of messages plus convenience helpers."""

    def __init__(self, messages: list[Message] | None = None) -> None:
        self.messages: list[Message] = messages or []

    def add(self, message: Message) -> None:
        self.messages.append(message)

    def add_assistant(self, turn: AssistantTurn) -> None:
        self.messages.append(
            Message(
                role=Role.ASSISTANT,
                text=turn.text,
                tool_calls=list(turn.tool_calls),
                raw=turn.raw,
            )
        )

    def add_tool_results(self, results: list[ToolResult]) -> None:
        self.messages.append(Message(role=Role.TOOL, tool_results=results))

    def system_text(self) -> str:
        return "\n\n".join(m.text for m in self.messages if m.role is Role.SYSTEM)

    def __len__(self) -> int:
        return len(self.messages)
